[PS] Implement Permission System

Cal-L opened the pull request:
**Development & PR Process**
1. Follow MetaMask [Mobile Coding Standards](https://docs.google.com/document/d/1VJLwTRsUw_5EDq_o8d6sSbXUAYENLurkRitYO45eY5o/edit?usp=sharing)
2. Add `release-xx` label to identify the PR slated for a upcoming release (will be used in release discussion)
3. Add `needs-dev-review` label when work is completed
4. Add `needs-qa` label when dev review is completed
5. Add `QA Passed` label when QA has signed off

**Description**

Implements Permission System, an account access management system.

**High level touch points**
- PermissionController
- Permission Utils
  - Permission specs get validated
  - Permission gets decorated with caveats when middleware is created
- Subjects/hosts now → future (snaps, sdk, walletconnect)
- Base - Account selection, AccountSelectorList, useAccounts hook, ENS optimization
- HOC - AccountConnect, AccountPermissions

**UI Changes**
- Bottom Navigation Bar with Wallet & Browser tabs
- Removal of Wallet & Browser from side drawer
- Browser URL bar with new UI & new network icon that toggles to wallet / Permissions Button once connected to the dapp.
- New Bottom Sheet where dapps accounts access can be controlled.
- Toast Notification on active account & network change

**Test Cases**
https://www.notion.so/Test-Cases-6c6726b6ff9e4c40b829ee0140b7997a

**Bug Report** 
https://github.com/MetaMask/mobile-planning/issues/528

**Screenshots/Recordings**

![Simulator Screen Recording - iPhone 13 - 2022-12-21 at 17 09 21](https://user-images.githubusercontent.com/1649425/208965999-007a83d6-c6bf-46db-95d8-f026289a4574.gif)

**Issue**

Check  #4754 for test cases, designs, knowledge shares & bug reports

**Checklist**

* [x] There is a related GitHub issue
* [x] Tests are included if applicable
* [x] Any added code is fully documented

Cal-L committed:
Feature/4761 update network picker on wallet screen (#4844)

* Fix image styling

* Update constant name

* Replace wallet navbar with PickerNetwork

Cal-L committed:
Feature/4768 action selector sheet (#4891)

* Create AccountSelectionList

* Create temp loader component

* Create temp sheet actions component

* Fix AvatarGroup positioning

* Provide reserved overlay spacing for bottom sheet

* Fix sheet header typo

* Make analytics params optional

* Rename AccountSelectorList

* Create new account selector

* Add AccountSelector to navigation stack

* Replace all account selection logic

* Improve caching logic for ENS name fetch. Introduce 1 hour cache to optimize for network calls. Fetching from middle out beginning with selected account.

* Add lodash debounce to prevent taps in quick succession on bottom sheet

* Throw away any debounce innvocation when bottom sheet is unmounted

* Introduce loading to account selector. Readd test IDS

* Remove old account list

* Clean up useAccounts

* Update snapshots

* Fix logic after pulling from main

* Update snapshots

Cal-L committed:
Add new logos (#5046)

Cal-L committed:
[PS] Fix component styles (#5047)

* Add new logos

* Update styles. Add missing exports.

Cal-L committed:
[PS] Patch controllers (#5048)

* Add new logos

* Update styles. Add missing exports.

* Patch controllers. Introduce react-native-get-random-values

Cal-L committed:
[PS] Organize account utils (#5049)

* Add new logos

* Update styles. Add missing exports.

* Patch controllers. Introduce react-native-get-random-values

* Organize account hooks and utils

* Export ENS cache. Update account util to return ENS name.

* Update account selector

* Remove account selector styles

Cal-L committed:
[PS] Hook up PS controller (#5050)

* Add new logos

* Update styles. Add missing exports.

* Patch controllers. Introduce react-native-get-random-values

* Organize account hooks and utils

* Export ENS cache. Update account util to return ENS name.

* Update account selector

* Remove account selector styles

* Hook up permission system controller logic

Cal-L committed:
[PS] Connect ps to browser (#5051)

* Add new logos

* Update styles. Add missing exports.

* Patch controllers. Introduce react-native-get-random-values

* Organize account hooks and utils

* Export ENS cache. Update account util to return ENS name.

* Update account selector

* Remove account selector styles

* Hook up permission system controller logic

* Connect PS to browser. Pass from address to transactions.

Cal-L committed:
[PS] Create first time connect sheet (#5052)

* Add new logos

* Update styles. Add missing exports.

* Patch controllers. Introduce react-native-get-random-values

* Organize account hooks and utils

* Export ENS cache. Update account util to return ENS name.

* Update account selector

* Remove account selector styles

* Hook up permission system controller logic

* Connect PS to browser. Pass from address to transactions.

* Create first time connect sheet

Cal-L committed:
[PS] Create permissions sheet (#5053)

* Add new logos

* Update styles. Add missing exports.

* Patch controllers. Introduce react-native-get-random-values

* Organize account hooks and utils

* Export ENS cache. Update account util to return ENS name.

* Update account selector

* Remove account selector styles

* Hook up permission system controller logic

* Connect PS to browser. Pass from address to transactions.

* Create first time connect sheet

* Create permissions sheet

jpcloureiro committed:
[PS] Tab bar bottom navigation (#4949)

* Add new logos

* Update styles. Add missing exports.

* Patch controllers. Introduce react-native-get-random-values

* Organize account hooks and utils

* Export ENS cache. Update account util to return ENS name.

* Update account selector

* Remove account selector styles

* Hook up permission system controller logic

* Connect PS to browser. Pass from address to transactions.

* Create first time connect sheet

* Create permissions sheet

* Add translations to permission system

* (BottomTabBar): implement new BottomTabBar

* (ActivityView): move nav stack to wallet

* (Transactions): update navigation header

* (ActivityView): fix back button onPress

* (ActivityView): fix getNavbarOptions args

* (BrowserBottomBar): remove bottomPadding

* (Transactions): fix empty screen height

* (Transactions): setup main modal stack navigator

Co-authored-by: Cal Leung <cleun007@gmail.com>

jpcloureiro committed:
[PS] Browser nav bar facelift (#4907)

* Add new logos

* Update styles. Add missing exports.

* Patch controllers. Introduce react-native-get-random-values

* Organize account hooks and utils

* Export ENS cache. Update account util to return ENS name.

* Update account selector

* Remove account selector styles

* Hook up permission system controller logic

* Connect PS to browser. Pass from address to transactions.

* Create first time connect sheet

* Create permissions sheet

* Add translations to permission system

* (BottomTabBar): implement new BottomTabBar

* (ActivityView): move nav stack to wallet

* (Transactions): update navigation header

* (ActivityView): fix back button onPress

* (ActivityView): fix getNavbarOptions args

* (BrowserBottomBar): remove bottomPadding

* (Transactions): fix empty screen height

* (Transactions): setup main modal stack navigator

* (BrowserNav): implement new PS browser navbar

* (Navbar): fix AccountRightButton onPress event

Co-authored-by: Cal Leung <cleun007@gmail.com>

Cal-L committed:
Feature/5068 code clean up cal (#5073)

* Feature/4768 action selector sheet (#4891)

* Create AccountSelectionList

* Create temp loader component

* Create temp sheet actions component

* Fix AvatarGroup positioning

* Provide reserved overlay spacing for bottom sheet

* Fix sheet header typo

* Make analytics params optional

* Rename AccountSelectorList

* Create new account selector

* Add AccountSelector to navigation stack

* Replace all account selection logic

* Improve caching logic for ENS name fetch. Introduce 1 hour cache to optimize for network calls. Fetching from middle out beginning with selected account.

* Add lodash debounce to prevent taps in quick succession on bottom sheet

* Throw away any debounce innvocation when bottom sheet is unmounted

* Introduce loading to account selector. Readd test IDS

* Remove old account list

* Clean up useAccounts

* Update snapshots

* Fix logic after pulling from main

* Update snapshots

* [PS] Connect ps to browser (#5051)

* Add new logos

* Update styles. Add missing exports.

* Patch controllers. Introduce react-native-get-random-values

* Organize account hooks and utils

* Export ENS cache. Update account util to return ENS name.

* Update account selector

* Remove account selector styles

* Hook up permission system controller logic

* Connect PS to browser. Pass from address to transactions.

* Update toast copy

* Fix misc

* Hide invalid fiat balance

* Comment permission util for clarity

* Move account type out of Toast

* Update imports

* Replace account depedency in browser tab

* Update styles

* Prevent auto account selection on account creation

* Remove spacing from toast translation strings

* Update browser icon

* Remove all permissions when deleting wallet

* Fix AvatarNetwork showfallback

* Open separate browser from warning

* Remove unnecessary padding on browser

* Show account switch toast when switching between tabs

* Fix unrecognized network logos

* Update snapshots

* Pull full screen modal/sheets to Root modal flow in nav stack

* Fix browser tab test

* Fix test

* Fix NetworkList test

* Remove redundant useAccounts

Cal-L committed:
Fix non-components tests

Cal-L committed:
Fix WalletConnect and transacting logic

jpcloureiro committed:
[PS] code cleanup (#5084)

* (Navigation): update route names

* (BrowserUrlBar): use getURLProtocol to check SSL

* (BrowserUrlBar): test margins

* (Browser): fix Android url bar onPress

Cal-L committed:
Update onboarding wizard

Cal-L committed:
don't show account toast when host name is the same

Cal-L committed:
Fix snapshot

joaoloureirop committed:
Merge branch 'main' into feature/4754-permission-system

joaoloureirop committed:
[PS]: update component library imports

joaoloureirop committed:
Merge branch 'main' into feature/4754-permission-system

joaoloureirop committed:
(CustomNetwork): update prop to allow wallet return

joaoloureirop committed:
Merge branch 'feature/5099-post-bug-bash-clean-up' into feature/4754-permission-system-bugbash-build

joaoloureirop committed:
update build number

joaoloureirop committed:
implement visual & ux updates

joaoloureirop committed:
(TestBuild) bump version

joaoloureirop committed:
Toast: respect account avatar global settings

joaoloureirop committed:
(TabBarItem): update component jest snapshot

joaoloureirop committed:
Merge branch 'main' into feature/4754-permission-system-bugbash-build

joaoloureirop committed:
(util/address): remove duplicate import

joaoloureirop committed:
Merge branch 'main' into feature/4754-permission-system-bugbash-build

Cal-L committed:
Merge from main

sethkfman committed:
updated android SDK versions

joaoloureirop committed:
update to react-native 0.66.5

more info https://github.com/facebook/react-native/issues/35210

andrepimenta committed:
SDK compatibility

andrepimenta committed:
Update coinbase

Cal-L committed:
Clean up SDK support a bit

Cal-L committed:
Add padding to TagURL

joaoloureirop committed:
fix linter

joaoloureirop committed:
Merge branch 'main' into feature/4754-permission-system

joaoloureirop committed:
fix typo

Cal-L committed:
Fix revoke button colors (#5255)

* Fix revoke button colors

* Remove unused import

Cal-L committed:
Hide tab bar on Android (#5261)

joaoloureirop committed:
Merge branch 'main' into feature/4754-permission-system

joaoloureirop committed:
fix lint errors

jpcloureiro committed:
update Permission System translations (#5401)

translations fetched from commit bda3e4a

jpcloureiro committed:
[PS] Allow Permission Request Rejection (#5405)

* get permission request id on AccountConnect

* refactor account connect action handlers

* move USER_INTENT enum to constants folder

* clenup

- remove eslint ignores
- remove old TODOs
- add useEffect comment to address why we need the user intent state
  machine

joaoloureirop committed:
(RootRPCMethodsUI): relocate requestData destruct

on Permission System we use a Permission Request ID that
is only relevant for the Permission Controller and not the other
Approvals.

Destructuring requestData to get said id should only happen on
REQUEST_PERMISSION Approval Type.

joaoloureirop committed:
Merge branch 'main' into feature/4754-permission-system

owencraston commented on app/component-library/components-temp/Loader/Loader.tsx:
```diff
@@ -0,0 +1,23 @@
+
+  return (
+    <View style={styles.base}>
+      <ActivityIndicator size={size} color={colors.primary.default} />
```
In some cases we might want to use the inverted colour if we are using the Primary button. See my [RestoreWallet screen](https://github.com/MetaMask/metamask-mobile/pull/4421/files#diff-fd04d31e4019fc6efd5c8a4b02f7404d583ae2c1b156f9990a9b195eaa7c5523R64) in the vault corruption PR as an example.

jpcloureiro replied:
addressed in b7c5f7c

owencraston commented on app/component-library/components-temp/SheetActions/SheetActions.styles.ts:
```diff
@@ -0,0 +1,28 @@
+    separator: {
+      height: 1,
+      backgroundColor: colors.border.muted,
+      marginHorizontal: 16,
```
should we use some standardized base value here? something like `marginHorizontal: space(2)`  or `marginHorizontal: 2* baseSpace,`,

jpcloureiro replied:
We should definitely standardize spacings across the project using Design System spacing scales.

I'll created an issue to handle that.

owencraston commented on app/component-library/components-temp/SheetActions/SheetActions.tsx:
```diff
@@ -0,0 +1,54 @@
+    [actions, styles.separator],
+  );
+
+  return <>{renderActions()}</>;
```
whats the reason for making this into as render function. from what I can see we only have one component that we are returning so i'd argue this render function just adds unneeded [indirection](https://en.wikipedia.org/wiki/Indirection#:~:text=In%20computer%20programming%2C%20indirection%20(also,value%20through%20its%20memory%20address.). Not a big issue though.

owencraston replied:
maybe there are benefits since its a useCallBack... 🤔

owencraston commented on app/component-library/components-temp/SheetActions/SheetActions.tsx:
```diff
@@ -0,0 +1,54 @@
+  return <>{renderActions()}</>;
+};
+
+export default SheetActions;
```
React.memo()

owencraston commented on app/component-library/components-temp/SheetActions/SheetActions.tsx:
```diff
@@ -0,0 +1,54 @@
+                  size={ButtonSize.Lg}
+                  disabled={disabled || isLoading}
+                  /* eslint-disable-next-line */
+                  style={{ opacity: disabled ? 0.5 : 1 }}
```
we should avoid inline styles.

jpcloureiro replied:
addressed in 9367b18

jpcloureiro committed:
move backup wallet reminder popup (#5402)

so it does not block the navigation bottom bar

owencraston commented on app/component-library/components-temp/SheetActions/SheetActions.tsx:
```diff
@@ -0,0 +1,54 @@
+  const renderActions = useCallback(
+    () =>
+      actions.map(
+        ({ label, onPress, testID, isLoading, disabled, variant }, index) => {
```
where do these values come from? Can we type them so its more clear?

jpcloureiro replied:
typed on `app/component-library/components-temp/SheetActions/SheetActions.types.ts`

owencraston commented on app/component-library/components/Avatars/AvatarGroup/AvatarGroup.tsx:
```diff
@@ -53,7 +56,7 @@ const AvatarGroup = ({ tokenList }: AvatarGroupProps) => {
   return (
     <View style={styles.base}>
-      <View style={styles.stack}>{renderTokenList}</View>
+      <View style={styles.stack}>{renderTokenList()}</View>
```
whats the rationale with using useCallback vs useMemo here? Seems like they achieve a similar result but useMemo seems more applicable.

owencraston replied:
we can also wrap the component in React.memo()

jpcloureiro replied:
Not sure if the component benefits of React.memo() alone. We would also need a comparator for the tokenList prop because React does shallow comparisons by default.
I don't see this component being that expensive to the point that we need to memoize it.

renderTokenList being a function instead of a value, I believe useCallback is better suited.

owencraston commented on app/component-library/components/Buttons/Button/foundation/ButtonBase/ButtonBase.styles.ts:
```diff
@@ -31,6 +31,7 @@ const styleSheet = (params: {
         alignItems: 'center',
         justifyContent: 'center',
         borderRadius: sizeAsNum / 2,
+        paddingHorizontal: 16,
```
same as above. should we define some global base units / function to generate these values for us?

owencraston commented on app/component-library/components/Cards/Card/Card.tsx:
```diff
@@ -15,9 +15,9 @@ const Card: React.FC<CardProps> = ({ style, children, ...props }) => {
 
   return (
-    <View style={styles.base} {...props}>
+    <TouchableOpacity style={styles.base} {...props}>
```
are we sure we want to make this entire component touchable? I don't see us enforcing an onPressHandler...what would happen here if we didn't handle the touch event? I think there would be some weird animations.

owencraston commented on app/component-library/components/Navigation/TabBar/TabBar.tsx:
```diff
@@ -21,6 +21,8 @@ const TabBar = ({ state, descriptors, navigation }: TabBarProps) => {
   const renderTabBarItem = useCallback(
     (route: { name: string; key: string }, index: number) => {
       const label = descriptors[route.key].options.tabBarLabel as TabBarLabel;
+      if (!label) return null;
```
should we completely return null from this component instead of returning a view with null in it? I think in this case the view will still be mounted

jpcloureiro replied:
addressed in cd002164

owencraston commented on app/component-library/components/Sheet/SheetBottom/SheetBottom.constants.ts:
```diff
@@ -17,4 +17,8 @@ export const SWIPE_TRIGGERED_ANIMATION_DURATION = 200;
  * The animation duration used for initial render.
  */
-export const INITIAL_RENDER_ANIMATION_DURATION = 500;
+export const INITIAL_RENDER_ANIMATION_DURATION = 350;
```
just curious why these were changed. Not a blocker though

owencraston commented on app/component-library/components/Sheet/SheetBottom/SheetBottom.tsx:
```diff
@@ -4,13 +4,14 @@
 } from 'react';
 import {
-  LayoutAnimation,
+  // LayoutAnimation,
```
🔥

owencraston commented on app/components/Nav/App/index.js:
```diff
@@ -347,15 +367,22 @@ const App = ({ userLoggedIn }) => {
+      screenOptions={clearStackNavigatorOptions}
+      initialRouteName={'DetectedTokens'}
     >
+      <Stack.Screen name={'DetectedTokens'} component={DetectedTokens} />
```
for route names you should use the Routes object.

owencraston commented on app/components/Nav/App/index.js:
```diff
@@ -365,10 +392,28 @@ const App = ({ userLoggedIn }) => {
         component={TurnOffRememberMeModal}
       />
+      <Stack.Screen
+        name={'AssetHideConfirmation'}
```
Routes

owencraston commented on app/components/Nav/App/index.js:
```diff
@@ -365,10 +392,28 @@ const App = ({ userLoggedIn }) => {
+        name={'AssetHideConfirmation'}
+        component={AssetHideConfirmation}
+      />
+      <Stack.Screen name={'DetectedTokens'} component={DetectedTokensFlow} />
```
😉

owencraston commented on app/components/Nav/Main/MainNavigator.js:
```diff
@@ -190,67 +164,93 @@ const WalletTabStackFlow = () => (
+  };
+
+  useEffect(() => {
+    // Hide keyboard on Android when keyboard is visible.
```
is there an issue to track this work?

owencraston commented on app/components/Nav/Main/MainNavigator.js:
```diff
@@ -616,9 +616,10 @@ const MainNavigator = () => (
       }}
     />
-    <Stack.Screen name="Home" tabBarVisible={false} component={HomeTabs} />
+    <Stack.Screen name="Home" component={HomeTabs} />
```
could we also add these to the routes file?

owencraston commented on app/components/Nav/Main/RootRPCMethodsUI.js:
```diff
@@ -705,7 +708,23 @@ const RootRPCMethodsUI = (props) => {
+              metadata: { id },
+            } = requestData;
+
+            props.navigation.navigate(Routes.MODAL.ROOT_MODAL_FLOW, {
```
I'd recommend using the `createNavigationDetails` pattern here. this will involve creating a function inside the `ACCOUNT_CONNECT` screen that defines the expected types. You can then call `props.navigation.navigate(...createAcountConnectNavDetails(....)` and you will get proper type checking on the params. This better enforces the correct params are used as well as allows someone to directly click into the AcountConnect screen as the function will be defined there. See [GetQuotes](https://github.com/MetaMask/metamask-mobile/blob/main/app/components/UI/FiatOnRampAggregator/Views/GetQuotes.tsx#L58-L67) as a reference

owencraston replied:
This can be applied to all navigate calls.

owencraston replied:
then you can use the [useParams](https://github.com/MetaMask/metamask-mobile/blob/main/app/components/UI/FiatOnRampAggregator/Views/GetQuotes.tsx#L219)  with proper destructuring and type checking.

owencraston commented on app/components/Nav/Main/index.js:
```diff
@@ -200,6 +214,38 @@ const Main = (props) => {
+    (state) => state.engine.backgroundState.NetworkController.provider,
+  );
+  const prevNetworkProvider = useRef(undefined);
+  const { toastRef } = useContext(ToastContext);
```
why are we using context for the Toast but redux for everything else?

owencraston commented on app/components/UI/AccountOverview/index.js:
```diff
@@ -222,17 +215,12 @@ class AccountOverview extends PureComponent {
+  openAccountSelector = () => {
+    const { onboardingWizard, navigation } = this.props;
+    !onboardingWizard &&
+      navigation.navigate(Routes.MODAL.ROOT_MODAL_FLOW, {
```
same comment as [this](https://github.com/MetaMask/metamask-mobile/pull/5062#discussion_r1054786938). lest try and use the `createNavigationDetails` pattern here.

owencraston commented on app/components/UI/AccountRightButton/index.tsx:
```diff
@@ -0,0 +1,96 @@
+import { BadgeVariants } from '../../../component-library/components/Badges/Badge/Badge.types';
+import BadgeWrapper from '../../../component-library/components/Badges/BadgeWrapper';
+
+const styles = StyleSheet.create({
```
should we move this to a separate file to better match the pattern in `component-library`?

owencraston commented on app/components/UI/AccountRightButton/index.tsx:
```diff
@@ -0,0 +1,96 @@
+  );
+};
+
+export default AccountRightButton;
```
React.memo() ?

owencraston commented on app/components/UI/AccountSelectorList/AccountSelectorList.tsx:
```diff
@@ -0,0 +1,224 @@
+  );
+
+  return (
+    <FlatList
```
thought on using a more performant list component like [FlashList](https://shopify.github.io/flash-list/) or [RecyclerListView](https://github.com/Flipkart/recyclerlistview)?

owencraston commented on app/components/UI/AccountSelectorList/AccountSelectorList.tsx:
```diff
@@ -0,0 +1,224 @@
+      keyExtractor={getKeyExtractor}
+      renderItem={renderAccountItem}
+      // Increasing number of items at initial render fixes scroll issue.
+      initialNumToRender={999}
```
this seems super hacky. what is going on here and is there a better way to solve the problem. Does this have performance implications?

owencraston commented on app/components/UI/AccountSelectorList/AccountSelectorList.tsx:
```diff
@@ -0,0 +1,224 @@
+        { cancelable: false },
+      );
+    },
+    /* eslint-disable-next-line */
```
is this needed?

owencraston commented on app/components/UI/AccountSelectorList/AccountSelectorList.tsx:
```diff
@@ -0,0 +1,224 @@
+  const getKeyExtractor = ({ address }: Account) => address;
+
+  const getTagLabel = (type: KeyringTypes) => {
+    let label = '';
```
since there are only two types here, can we turn this into a ternary so we guarantee label is not an empty string?

owencraston commented on app/components/UI/AccountSelectorList/AccountSelectorList.tsx:
```diff
@@ -0,0 +1,224 @@
+        });
+      }, 0);
+    }
+    // eslint-disable-next-line
```
is this needed?

owencraston commented on app/components/UI/AccountSelectorList/AccountSelectorList.tsx:
```diff
@@ -0,0 +1,224 @@
+    );
+    if (account) {
+      // Wrap in timeout to provide more time for the list to render.
+      setTimeout(() => {
```
I get bad vibes from setTimeout. is there something else we can wait on to give the list more time to render? Maybe using a more performant list implementation as I mentioned before would render this useless?

owencraston commented on app/components/UI/AccountSelectorList/AccountSelectorList.tsx:
```diff
@@ -0,0 +1,224 @@
+    return label;
+  };
+
+  const renderAccountBalances = useCallback(
```
Nit Pick. I think this is more of a case for useMemo.

owencraston commented on app/components/UI/AccountSelectorList/AccountSelectorList.tsx:
```diff
@@ -0,0 +1,224 @@
+
+      return (
+        <Cell
+          onLongPress={() => {
```
is there a way to pass this in directly without the `() => {}`. I may be wrong but I think doing it this way will cause a full re render every time the component is mounted.

owencraston commented on app/components/UI/AccountSelectorList/AccountSelectorList.tsx:
```diff
@@ -0,0 +1,224 @@
+          title={accountName}
+          secondaryText={shortAddress}
+          tertiaryText={balanceError}
+          onPress={() => onSelectAccount?.(address, isSelectedAccount)}
```
same as above

owencraston commented on app/components/UI/AccountSelectorList/AccountSelectorList.tsx:
```diff
@@ -0,0 +1,224 @@
+          secondaryText={shortAddress}
+          tertiaryText={balanceError}
+          onPress={() => onSelectAccount?.(address, isSelectedAccount)}
+          avatarProps={{
```
should we create a separate object for this so this prop can be cached by React?

owencraston commented on app/components/UI/AccountSelectorList/AccountSelectorList.tsx:
```diff
@@ -0,0 +1,224 @@
+          tagLabel={tagLabel}
+          disabled={isDisabled}
+          /* eslint-disable-next-line */
+          style={{ opacity: isLoading ? 0.5 : 1 }}
```
I think we should avoid inline styles for the same reason mentioned above. this guarantees a full re render on mount. we can simply assign this object and the problem will be fixed.

jpcloureiro replied:
addressed in 9367b185

owencraston commented on app/components/UI/AccountSelectorList/AccountSelectorList.tsx:
```diff
@@ -0,0 +1,224 @@
+          style={{ opacity: isLoading ? 0.5 : 1 }}
+        >
+          {renderRightAccessory?.(address, accountName) ||
+            (assets && renderAccountBalances(assets))}
```
nit pick: I prefer to use a ternary and explicitly return null in these cases for the [following reason](https://kentcdodds.com/blog/use-ternaries-rather-than-and-and-in-jsx).

owencraston commented on app/components/UI/AccountSelectorList/AccountSelectorList.tsx:
```diff
@@ -0,0 +1,224 @@
+  );
+};
+
+export default AccountSelectorList;
```
should these be wrapped in a React.memo()? I am not sure here in the case that is will be re rendered a lot but if not we should try to memoize it.

owencraston commented on app/components/UI/ApproveTransactionReview/index.js:
```diff
@@ -803,7 +794,7 @@ class ApproveTransactionReview extends PureComponent {
     InteractionManager.runAfterInteractions(() => {
       this.onCancelPress();
-      this.props.navigation.navigate(Routes.BROWSER_VIEW, {
+      this.props.navigation.navigate(Routes.BROWSER.VIEW, {
```
can we adopt the `createNavigationDetails` pattern here. See other [comments](https://github.com/MetaMask/metamask-mobile/pull/5062#discussion_r1054786938) as reference.

owencraston commented on app/components/UI/AssetOverview/index.js:
```diff
@@ -223,11 +223,10 @@ class AssetOverview extends PureComponent {
   goToBrowserUrl(url) {
-    this.props.navigation.navigate(Routes.BROWSER_TAB_HOME, {
-      screen: Routes.BROWSER_VIEW,
+    this.props.navigation.navigate('Webview', {
```
see comments about [createNavDetails](https://github.com/MetaMask/metamask-mobile/pull/5062#discussion_r1054786938). Also why is this changing to a SimpleWebView? We should add this to routes as well.

jpcloureiro replied:
addressed in 572425d2
About the webview, we are opening here a documentation website. It seems more appropriate to open it on a simple webview instead of the in-app browser.

owencraston commented on app/components/UI/BrowserUrlBar/BrowserUrlBar.tsx:
```diff
@@ -0,0 +1,62 @@
+
+    if (
+      isGatewayUrl(urlObj) &&
+      url.search(`${AppConstants.IPFS_OVERRIDE_PARAM}=false`) === -1 &&
```
what's this for?

owencraston replied:
does this have performance implications?

owencraston commented on app/components/UI/BrowserUrlBar/BrowserUrlBar.tsx:
```diff
@@ -0,0 +1,62 @@
+    if (!url) return;
+
+    const urlObj = new URL(url);
+    const ensUrl = route.params?.currentEnsName ?? '';
```
if we leveraged the `useParams` hook here we could get better typing.

owencraston commented on app/components/UI/BrowserUrlBar/BrowserUrlBar.tsx:
```diff
@@ -0,0 +1,62 @@
+    return urlObj.host.toLowerCase().replace(/^www\./, '');
+  };
+
+  const contentProtocol = getURLProtocol(url);
```
these values could be memoized

owencraston commented on app/components/UI/DrawerView/index.js:
```diff
@@ -618,16 +607,17 @@ class DrawerView extends PureComponent {
-      this.trackEvent(ANALYTICS_EVENT_OPTS.NAVIGATION_TAPS_ACCOUNT_NAME);
+  openAccountSelector = () => {
+    const { navigation } = this.props;
+    navigation.navigate(Routes.MODAL.ROOT_MODAL_FLOW, {
```
[createNavDetails](https://github.com/MetaMask/metamask-mobile/pull/5062#discussion_r1054786938)

jpcloureiro replied:
addressed in 572425d

owencraston commented on app/components/UI/DrawerView/index.js:
```diff
@@ -718,15 +709,18 @@ class DrawerView extends PureComponent {
+    // Q: duplicated analytic event?
     this.trackOpenBrowserEvent();
+    // NOTE: do we need this event?
     this.trackEvent(ANALYTICS_EVENT_OPTS.NAVIGATION_TAPS_BROWSER);
```
@gantunesr can we remove this legacy event or are both needed? same goes for the below event.

owencraston commented on app/components/UI/DrawerView/index.js:
```diff
@@ -1426,28 +1380,6 @@ class DrawerView extends PureComponent {
             onClose={this.closeInvalidCustomNetworkAlert}
           />
         </Modal>
-        <Modal
```
I can't tell you how happy it makes me to see lines being deleted from this file?

owencraston commented on app/components/UI/FiatOnRampAggregator/components/AccountSelector.tsx:
```diff
@@ -37,11 +38,13 @@ const AccountSelector = () => {
-    dispatch(toggleAccountsModal());
-  }, [dispatch]);
+  const openAccountSelector = () =>
+    navigation.navigate(Routes.MODAL.ROOT_MODAL_FLOW, {
```
[createNavDetails](https://github.com/MetaMask/metamask-mobile/pull/5062/files#)

jpcloureiro replied:
addressed in 572425d

owencraston commented on app/components/UI/FiatOrders/components/AccountSelector.js:
```diff
@@ -19,22 +20,26 @@ const styles = StyleSheet.create({
+  const navigation = useNavigation();
+
+  const openAccountSelector = () =>
+    navigation.navigate(Routes.MODAL.ROOT_MODAL_FLOW, {
```
[createNavDetails](https://github.com/MetaMask/metamask-mobile/pull/5062/files#). This could also be a useCallback

jpcloureiro replied:
addressed in 572425d

owencraston commented on app/components/UI/FiatOrders/components/AccountSelector.js:
```diff
@@ -19,22 +20,26 @@ const styles = StyleSheet.create({
+
+  return (
+    <SelectorButton onPress={openAccountSelector} style={styles.selector}>
+      <Identicon diameter={15} address={selectedAddress} />
```
is there a standard value we can use here?

jpcloureiro replied:
Identicon diameter prop value wasn't changed

owencraston commented on app/components/UI/NetworkModal/index.tsx:
```diff
@@ -196,7 +198,9 @@ const NetworkModals = (props: NetworkProps) => {
     NetworkController.setRpcTarget(url.href, decimalChainId, ticker, nickname);
     closeModal();
-    navigation.navigate('WalletView');
+    shouldNetworkSwitchPopToWallet
```
is there a Route value for wallet view?

owencraston commented on app/components/UI/OnboardingWizard/Coachmark/index.js:
```diff
@@ -99,6 +99,12 @@ const createStyles = (colors) =>
     },
+    bottomRight: {
+      marginBottom: 10,
+      top: -2,
```
why -2 here?

owencraston commented on app/components/UI/OnboardingWizard/Step5/index.js:
```diff
@@ -30,49 +33,33 @@ const styles = StyleSheet.create({
     navigation &&
-      navigation.navigate(Routes.BROWSER_TAB_HOME, {
-        screen: Routes.BROWSER_VIEW,
+      navigation.navigate(Routes.BROWSER.HOME, {
```
[createNavDetails](https://github.com/MetaMask/metamask-mobile/pull/5062#discussion_r1054786938)

jpcloureiro replied:
addressed in 572425d

jpcloureiro replied:
addressed in 572425d

owencraston commented on app/components/UI/OnboardingWizard/Step6/index.js:
```diff
@@ -106,6 +105,10 @@ const mapDispatchToProps = (dispatch) => ({
+  /**
+   * Object that represents the navigator
+   */
+  navigation: PropTypes.object,
```
can we not call useNavigation in this file?

owencraston commented on app/components/UI/Tokens/index.js:
```diff
@@ -323,7 +324,9 @@ class Tokens extends PureComponent {
     const { NetworkController } = Engine.context;
     const { detectedTokens } = this.props;
-    this.props.navigation.navigate('DetectedTokens');
+    this.props.navigation.navigate(Routes.MODAL.ROOT_MODAL_FLOW, {
```
[createNavDetails](https://github.com/MetaMask/metamask-mobile/pull/5062#discussion_r1054786938)

jpcloureiro replied:
addressed in 572425d

jpcloureiro replied:
addressed in 572425d

owencraston commented on app/components/UI/Tokens/index.js:
```diff
@@ -323,7 +324,9 @@ class Tokens extends PureComponent {
     const { detectedTokens } = this.props;
-    this.props.navigation.navigate('DetectedTokens');
+    this.props.navigation.navigate(Routes.MODAL.ROOT_MODAL_FLOW, {
+      screen: 'DetectedTokens',
```
detect Tokens could be in Routes

jpcloureiro replied:
addressed in 572425d

owencraston commented on app/components/UI/TransactionReview/TransactionReviewInformation/index.js:
```diff
@@ -511,7 +511,7 @@ class TransactionReviewInformation extends PureComponent {
     InteractionManager.runAfterInteractions(() => {
       this.onCancelPress();
-      this.props.navigation.navigate(Routes.BROWSER_VIEW, {
+      this.props.navigation.navigate(Routes.BROWSER.VIEW, {
```
[createNavDetails](https://github.com/MetaMask/metamask-mobile/pull/5062#discussion_r1054786938)

jpcloureiro replied:
addressed in 572425d

jpcloureiro replied:
addressed in 572425d

jpcloureiro commented on app/component-library/components/Avatars/AvatarGroup/AvatarGroup.types.ts:
```diff
@@ -30,3 +26,11 @@ export interface AvatarGroupProps extends ViewProps {
+/**
+ * Style sheet input parameters.
+ */
+export interface AvatarGroupStyleSheetVars {
```
These values are used to calculate container height & width based on the amount of avatars.
They are not component properties but stylesheet parameters.

Cal-L committed:
Fix bug report 12 & 3

Cal-L committed:
Merge from main

Cal-L committed:
Use MetaMetrics and update snapshots

joaoloureirop committed:
(Loader): add color prop

joaoloureirop committed:
remove inline styles

joaoloureirop committed:
(TabBar): remove null label condition.

This condition is not needed anymore since we removed
the Activity Route from the bottom navigation stack

Cal-L committed:
Fix snapshots

Cal-L committed:
Merge branch 'feature/4754-permission-system' of https://github.com/MetaMask/metamask-mobile into feature/4754-permission-system

Cal-L committed:
Merge with main

Cal-L committed:
Improve logic to autoscroll account selector

Cal-L committed:
Fix analytics

Cal-L committed:
Enable ability to delete imported account while in connect mode

Cal-L committed:
Auto scroll multiselect when possible

Cal-L committed:
Fix lint

Cal-L committed:
Merge with main

Cal-L committed:
Replace @metamask/controllers imports

joaoloureirop committed:
(Navigation): adopt useNavigationDetails pattern

Cal-L committed:
Improve wallet tab contrast

Cal-L committed:
Merge branch 'feature/4754-permission-system' of https://github.com/MetaMask/metamask-mobile into feature/4754-permission-system

Cal-L committed:
Remove unused network prop from browser tab

joaoloureirop committed:
force permitted accounts to be lower case

Cal-L committed:
Change favicon size to 50. Opensea doesn't support 64

cortisiko commented:
bug report: https://github.com/MetaMask/mobile-planning/issues/528

Cal-L committed:
Update revoke translations

Cal-L committed:
Merge with main

Cal-L committed:
Fix lint issue

Cal-L committed:
Fix missing account name in toast

Cal-L committed:
Only allow removing imported wallets via seed

Cal-L committed:
Merge in main

Cal-L committed:
Improve avatar network placeholder padding

Cal-L committed:
Improve network icon display when switching networks

cortisiko committed:
E2e/permission system (#5539)

* fix first set of broken tests

* update tokens text

* fix deep linking assertion

Cal-L committed:
Merge with main

cortisiko committed:
fix snapshot

Cal-L committed:
Fix storybook

Cal-L committed:
Merge branch 'feature/4754-permission-system' of https://github.com/MetaMask/metamask-mobile into feature/4754-permission-system

Cal-L committed:
Fix XS avatar alignment

Cal-L committed:
Lower font size for XS avatar fallback

Cal-L committed:
Use test network images

Cal-L committed:
Keep import flow persistent within account selection

Cal-L committed:
Reduce picker network to 1 line

Cal-L committed:
Re-add skip button for onboarding wizard

Cal-L committed:
Clean up network icon on network info screen when switching networks

Cal-L committed:
Show toast when switching between networks with same chainId but different type

cortisiko committed:
fix onboarding tutorial appium test

cortisiko committed:
fix appium tests

cortisiko committed:
forgot a commit to the helpers file.

joaoloureirop committed:
Merge branch 'main' into feature/4754-permission-system

Cal-L committed:
Merge from mina

cortisiko committed:
fix broken wallet.steps

Cal-L committed:
Handle adding and removing accounts during connect flows

Cal-L committed:
Merge branch 'feature/4754-permission-system' of https://github.com/MetaMask/metamask-mobile into feature/4754-permission-system

Cal-L committed:
Enable passing interacting address to watch asset

cortisiko committed:
fix detox wallet e2e

jpcloureiro committed:
Permission System UI bugfixes (#5603)

* fix network casing name in educational network info modal

* fix no network avatar for frequent networks

SamuelSalas committed:
E2e appium/fix browser test permission system (#5611)

* updated browser tests to run on thje permission system

* Returning values to original state

* Solve account approval modal issues and option menu test issues

* Update network text value

* updating failing tests

* Merge latest changes an returning config values to default

* Update

* Add AddContact.js file

* update AddContact page object + fix linting issue

---------

Co-authored-by: Curtis <Curtis.David7@gmail.com>

cortisiko committed:
Fix E2E merge conflicts on Permission System Branch (#5635)

* [IMPROVEMENT] Refactor ProtectYourWallet section in the security section (#5600)

* [FIX] Analytics event interface (#5613)

* Fix analytics events

* Bump version and build numbers

* Revert "Bump version and build numbers"

This reverts commit 247789b8428ad7f3bd3cca858c7818ffc5a7a68a.

* Handle ENS Address Error (#5605)

* resolving ens address

* lint

* cleaner name

* Bump versions to make prod build for testing

* Revert "Bump versions to make prod build for testing"

This reverts commit 43d0dad69582f5b3d467e4a37e9dd118cb36016a.

Co-authored-by: Plasma Corral <32695229+plasmacorral@users.noreply.github.com>

* [FIX] Make Prettier work in wdio directory (#5612)

* make Prettier work in wdio directory

- remove wdio from the Prettier exclude file
- add Gherkin Prettier pluggin for .feature
- update `yarn format` script to take .feature files into account
- first pass of js and feature files formating in wdio dir using `yarn format`

* format after branch update from main

* extract eslint config to deal with it in another PR

- create an eslint ignore file and use it as default instead of prettier ignore with eslint (on both scripts, lint and fix)
- add the wdio dir back in .eslintignore to deal with it in another eslint only PR
- remove useless ignore lines as already ignored by the android line

* [FIX] SRP Quiz content and translations (#5626)

* [IMPROVEMENT] Refactor Personal Signature (#5551)

* personal signature refactor

* Fix currency display (#5623)

* [FEAT] Add friction to SRP reveal (#5579)

* Add base modals

* Add navigation for dev purposes

* Update Main Nav

* Update navigation

* Add new Route

* Update quiz files

* Update modals

* Add QuizStage enum

* Add interfaces

* Update styles

* Update quiz components

* Update IQuizInformationProps interface

* Rename files

* Add quiz flow

* Revert AccountView changes

* Add Quiz to security settings

* Update title interface and style

* Update QuizContent

* Update styles

* Update dep arrays

* Add quiz structure

* Move QuizStage to type file

* Update QuizContent component

* Add method to dismiss modal

* Revert Podfile.lock

* Fix typo

* Remove unnecessary comments

* Fix structure error

* Add analytic events

* Fix small content issues

* Add image to quiz intro

* Add image to quiz intro

* Add link to zendesk article

* Remove quiz structure cobject

* Bump version and build number

* Ignore advisories

* Revert "Bump version and build number"

This reverts commit b9833bb2db5430d99985c3db269cbc30b695746f.

* Revert "Ignore advisories"

This reverts commit 17d68ebf7ba9b684425226eae3225e2c8631e25e.

* Remove message

* Update dependencies array

* Use Routes for RevealPrivateCredentialView

* Remove unused interfaces

* Move SRP guide method to QuizModal

* Update dependency array

* Refactor switch-case statement

* Fix nav bug

* Dismiss Quiz if the app goes to the background

* E2e appium folder structure change (#5632)

* Re-arranged wdio file structure

Re-arranged wdio file structure & refactor imports

* Added missing Address Input ID

* Commented out failing test

* Removed ignored test

* rename folder to "Accounts"

* move lockResetPassword out of onboarding folder

---------

Co-authored-by: Curtis <Curtis.David7@gmail.com>

* fix import path

* fix e2e after merging main.

* fix linting

* Remove unused files. Fix personal sign

---------

Co-authored-by: Gustavo Antunes <17601467+gantunesr@users.noreply.github.com>
Co-authored-by: Sylva Elendu <iamsylvaelendu@gmail.com>
Co-authored-by: Plasma Corral <32695229+plasmacorral@users.noreply.github.com>
Co-authored-by: Nicolas MASSART <NicolasMassart@users.noreply.github.com>
Co-authored-by: Cal Leung <cleun007@gmail.com>
Co-authored-by: yande <110056475+Andepande@users.noreply.github.com>

Cal-L committed:
Fix/4754 importing wallet via qr (#5634)

* Change ImportPrivateKey to TS file

* Add QR scanner screen to import stack

* Change ts to tsx

* Refactor import private key

* Don't show alert when importing account from account selector

* Fix deleting last imported account when in connected flow

* Fix dismissing keyboard when selecting top right nav button on webview

* Fix contrast for private key QR code

* Disable Permission System for WalletConnect and MMSDK

* Fix undefined origin issue

* fix browser title bar misalignment

* fix networkinfo avatar letter

* Resolve PR comments

---------

Co-authored-by: João Loureiro <joao.loureiro@consensys.net>

Cal-L committed:
Merge with main

jpcloureiro committed:
Add analytics to permission system (#5500)

* place analytics events

* rename permission connect request events

add source property

* update analytics

* fix import

* safeguarding when there are no active tabs

* rename event properties

* update source field

* update request completed properties

* fix networklist test

Cal-L committed:
Fix lint issues

cortisiko committed:
temporary bandaid for elements that do not appear.

cortisiko reviewed (approved):
This feature is QA passed 🌮 🌮 . The e2e tests were updated to reflect changes in this feature.  they all passed. See [iOS test run ](https://app.bitrise.io/build/26f5ac61-3028-4bf3-b30a-3d94ba16caf4) and [android test run](https://www.screencast.com/t/GgNjx80uEgVj). 

Test coverage:
- See [my feature file](https://www.notion.so/Permission-system-Feature-File-cbeadfd6d7f14538acfe9ebc8286780a) with most of my test scenarios for this feature. 
- Removing networks while connected to a dapp and ensuring the network defaults back to mainnet. I also ensured transactions were successful after network removal. 
- dapp and regular transactions (send, swap, approve tokens, deploy contracts, mint nfts)

owencraston commented on patches/@metamask+keyring-controller+1.0.1.patch:
```diff
@@ -0,0 +1,80 @@
+-            this.setSelectedAddress(accounts[0]);
+-            return this.fullUpdate();
++            const importedAccountAddress = accounts[0];
++            return { importedAccountAddress, keyringState: this.fullUpdate() };
```
Follow up with fix to controllers

gantunesr replied:
+1 this patch makes the development of Ledger more difficult because we're using a preview version of the keyring-controller

owencraston commented on patches/@metamask+keyring-controller+1.0.0.patch:
```diff
@@ -1,39 +0,0 @@
--            this.setSelectedAddress(accounts[0]);
--            return this.fullUpdate();
-+            const importedAccountAddress = accounts[0];
-+            return { importedAccountAddress, keyringState: this.fullUpdate() };
```
Follow up with fix to controllers

owencraston commented on patches/@metamask+assets-controllers+1.0.1.patch:
```diff
@@ -0,0 +1,129 @@
++     * @param params.token.image - Image of the token.
++     * @returns Current token list.
++     */
++    addTokenToAccount({ accountAddress, token: { address, symbol, decimals, image } }) {
```
Follow up with fix to controllers

owencraston commented on app/core/BackgroundBridge/BackgroundBridge.js:
```diff
@@ -337,13 +340,17 @@ export class BackgroundBridge extends EventEmitter {
-      }),
-    );
+    // TODO - Remove this condition when WalletConnect and MMSDK uses Permission System.
+    if (this.isMMSDK || this.isWalletConnect) {
```
you could just only run this when both of these are false.

owencraston committed:
remove useless else

owencraston reviewed (approved):
Reviewed latest changes with @Cal-L and it all looks good to me. Just left a few comments for future improvements.
